Games/Solitaire/SolitaireWidget.cpp

```diff
@@ -0,0 +1,424 @@
+                    } else if (m_focused_stack->type() == CardStack::Type::Normal
+                        && stack.type() == CardStack::Type::Foundation) {
+                        update_score(10);
+                    } else if (m_focused_stack->type() == CardStack::Type::Foundation
+                        && stack.type() == CardStack::Type::Normal) {
+                        update_score(-15);
+                    }
+
+                    rebound = false;
+                    break;
+                }
+            }
+        }
+    }
+
+    if (rebound) {
+        for (auto& to_intersect : m_focused_cards)
+            intersect_card(to_intersect);
+
+        m_focused_stack->rebound_cards();
+        m_focused_stack->set_dirty();
+    }
+
+    m_mouse_down = false;
+    m_has_to_repaint = true;
+}
+
+void SolitaireWidget::mousemove_event(GUI::MouseEvent& evt)
+{
+    GUI::Widget::mousemove_event(evt);
+
+    if (!m_mouse_down || m_game_over_animation)
+        return;
+
+    auto click_location = evt.position();
+    int dx = click_location.dx_relative_to(m_mouse_down_location);
+    int dy = click_location.dy_relative_to(m_mouse_down_location);
+
+    for (auto& to_intersect : m_focused_cards) {
+        intersect_card(to_intersect);
+        to_intersect.move(dx, dy);
+    }
+
+    m_mouse_down_location = evt.position();
+    m_has_to_repaint = true;
+}
+
+void SolitaireWidget::doubleclick_event(GUI::MouseEvent& evt)
+{
+    GUI::Widget::doubleclick_event(evt);
+
+    if (m_game_over_animation) {
+        enable_game_over_animation(false);
+        setup();
+        return;
+    }
+
+    auto click_location = evt.position();
+
+    for (auto& to_check : m_stacks) {
+        if (to_check.type() == CardStack::Type::Foundation)
+            continue;
+
+        if (to_check.bounding_box().contains(click_location) && !to_check.is_empty()) {
+            auto& top_card = to_check.peek();
+            if (!top_card.is_upside_down() && top_card.rect().contains(click_location)) {
+                if (stack(Foundation1).check_move(top_card))
+                    move_card(to_check, stack(Foundation1));
+                else if (stack(Foundation2).check_move(top_card))
+                    move_card(to_check, stack(Foundation2));
+                else if (stack(Foundation3).check_move(top_card))
+                    move_card(to_check, stack(Foundation3));
+                else if (stack(Foundation4).check_move(top_card))
+                    move_card(to_check, stack(Foundation4));
+                else
+                    break;
+
+                update_score(10);
+            }
+            break;
+        }
+    }
+
+    m_has_to_repaint = true;
+}
+
+void SolitaireWidget::check_win()
+{
+    for (auto& stack : m_stacks) {
+        if (stack.type() != CardStack::Type::Foundation)
+            continue;
+        if (stack.count() != Card::s_card_count)
+            return;
+    }
+
+    enable_game_over_animation(true);
```

Comment:
The method name reads to me as if it toggles whether there would be an animation shown on game over (like maybe it's a preference the user would set). But looking at how it's used; it actually start and stops the actual animation right at the time of the call? Then maybe it should be named `start/hide/show_game_over_animation()`